### ISSUE TITLE
Tensorflow 2.0 rejection resample not producing the proper target distribution

### ISSUE THREAD

**Issue by celinew1221** [tensorflow/tensorflow#28598]
**System information**
- Have I written custom code: Yes
- OS Platform and Distribution: Linux Ubuntu 16.04
- TensorFlow installed from (source or binary): from binary pip3
- TensorFlow version (use command below): 2.0-alpha
- Python version: 3.5.2
- CUDA/cuDNN version: 10.0
- GPU model and memory: GTX Titan V 12GB

**Issues**
I am trying to test rejection sampling with the following code. 
This is the result I got:

    target_dist [0.5, 0.5] 
    initial distribution [0.8333333333333334, 0.16666666666666666]
    result counts [1500, 600] 
    final dist 0.7142857142857143 0.2857142857142857

**Expectation**
The final distribution does not reflect the target distribution I set. I feel like this is a bug in the original algorithm because I can't seem to find what I did wrong in my code as shown below:

```
import tensorflow as tf
import numpy as np
# everything is based on tensorflow 2.0
tf.random.set_seed(2342)


def map2label(sample):
    return tf.cast(tf.math.equal(sample, 2), tf.int32)

np_data = np.array([0,2,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2])
target_dist = [0.5, 0.5]
init_dist = [(np_data.shape[0]-3)/np_data.shape[0], 3/np_data.shape[0]]

dataset = tf.data.Dataset.from_tensor_slices(np_data)
rej = tf.data.experimental.rejection_resample(map2label, target_dist, init_dist, 2342)   # set seed explicitly
dataset = dataset.apply(rej)

bucket_counts = [0, 0]
for i in range(100):
    for data in dataset:
        class_id, data_content = data
        bucket_counts[class_id.numpy()] += 1

print("This is your target_dist", target_dist, "This is your initial distribution", init_dist)
print("This is your result counts", bucket_counts,
      "This is your final dist", bucket_counts[0] / np.sum(bucket_counts), bucket_counts[1] / np.sum(bucket_counts))

```

**Comment by muddham** [comment]
@celinew1221 It looks like you haven't used a template to create this issue. Please resubmit your issue using a template from [here](https://github.com/tensorflow/tensorflow/issues/new/choose). We ask users to use the template because it reduces overall time to resolve a new issue by avoiding extra communication to get to the root of the issue. We will close this issue in lieu of the new one you will create from the template. Thank you for your cooperation.

**Comment by celinew1221** [comment]
Sure. Can I just edit this post? Looks like I am missing some system info.

**Comment by muddham** [comment]
@celinew1221 Able to reproduce the issue with the provided code. In the template it is mentioned that it is not a custom written code, can you provide the link for this, to investigate further.

This is your target_dist [0.5, 0.5] This is your initial distribution [0.8333333333333334, 0.16666666666666666]
This is your result counts [1500, 600] This is your final dist 0.7142857142857143 0.2857142857142857

**Comment by celinew1221** [comment]
Sorry, I meant that I didn't create a custom made rejection_resample function. Everything else is what I wrote. I'll edit that.

**Comment by shishirprao** [comment]
Hi,
I had a similar issue. Initially, I was using shuffle and repeat separately. This caused the rejection resample to produce incorrect distribution.  

Code
```python 
            self.data = self.data.shuffle(buffer_size=len(self.images)*2, seed= 1)
            self.data = self.data.repeat(count = 1)
            target_dist=[1.0/self.args.num_classes for cc in range(self.args.num_classes)]
            rej = tf.data.experimental.rejection_resample(
                          class_func=lambda _,__,___, y,____,_____:y,
                          initial_dist=self.class_weight.astype(np.float32),
                          target_dist=target_dist)
            self.data = self.data.apply(rej).map(lambda _,b:(b))
```
The incorrect result is provided below:
```result
With seperate shuffle and repeat
[0 0 3 3 3 0 1 0 3 0 3 0 0 0 0 0]
```

However, when I replaced the shuffle and repeat using the function given below, it works as expected.

```python
self.data = self.data.apply(tf.data.experimental.shuffle_and_repeat(buffer_size=len(self.images)*2,count=1))
```
Result with correct distribution is given below:
```result
With combined shuffle and repeat
[3 1 3 1 3 3 0 1 3 0 1 4 2 4 3 0]
```

**Comment by gadagashwini** [comment]
Issue is replicating with `Tf-nightly==2.2.0.dev20200311`,
Please find the gist [here](https://colab.sandbox.google.com/gist/gadagashwini/0501e4610fd758d4ba6e72a95d35604b/untitled443.ipynb). Thanks!

**Comment by jsimsa** [comment]
This issue should be fixed by https://github.com/tensorflow/tensorflow/commit/f313fdce45d4933938a89980f6ca9bb2c8cbd27a

**Comment by tloki** [comment]
Well, I'm stil able to reproduce the issue by running the gist @gadagashwini provided using 2.2.0-rc3, and I'm also having a hard time getting it to properly resample using my own training scripts (which I don't think is important to provide since even for a simple script from that gist - it does not work).

**Comment by Hoda1394** [comment]
I still have this issue with my own data (TensorFlow 2.2) and I can reproduce the gist that is provided above even by version 2.4.0-dev20200717! it seems that the problem still exists.
I think the issue needs to be reopened. @jsimsa

**Comment by TorstenK91** [comment]
I can reproduce the same gist using Tensorflow 2.4.3. This is very likely not fixed yet!
